cvmfs/ingestion/ingestion_source.h

```diff
@@ -0,0 +1,117 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_INGESTION_INGESTION_SOURCE_H_
+#define CVMFS_INGESTION_INGESTION_SOURCE_H_
+
+#include <pthread.h>
+
+#include <cerrno>
+#include <cstdio>
+#include <string>
+
+#include "duplex_libarchive.h"
+#include "platform.h"
+#include "util/posix.h"
+#include "util/single_copy.h"
+#include "util_concurrency.h"
+
+class IngestionSource : SingleCopy {
+ public:
+  virtual ~IngestionSource() {}
+  virtual std::string GetPath() const = 0;
+  virtual bool Open() = 0;
+  virtual ssize_t Read(void* buffer, size_t nbyte) = 0;
+  virtual bool Close() = 0;
+  virtual bool GetSize(uint64_t* size) = 0;
+};
+
+class FileIngestionSource : public IngestionSource {
+ public:
+  explicit FileIngestionSource(const std::string& path)
+      : path_(path), fd_(-1) {}
+  ~FileIngestionSource() {}
+
+  std::string GetPath() const { return path_; }
+
+  bool Open() {
+    fd_ = open(path_.c_str(), O_RDONLY);
+    if (fd_ < 0) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+               "Err: Impossible to open the file: %s (%d)\n %s", path_.c_str(),
+               errno, strerror(errno));
+      return false;
+    }
+    return true;
+  }
+
+  ssize_t Read(void* buffer, size_t nbyte) {
+    assert(fd_ >= 0);
+    ssize_t read = SafeRead(fd_, buffer, nbyte);
+    if (read < 0) {
```

Comment:
@jblomer I have mixed opinions here. From one side I do agree that one check could be enough, on the other hand, I do prefer to check also here since it is more specific and can provide better information about something that went wrong. In either case, something should be done, either remove this log or adding an extra one in the relative tar method.